Source/WebCore/page/Navigation.cpp

```diff
@@ -717,48 +762,72 @@ bool Navigation::innerDispatchNavigateEvent(NavigationNavigationType navigationT
-            // If a new event has been dispatched in our event handler then we were aborted above.
-            if (m_ongoingNavigateEvent != event.ptr())
-                return false;
+        waitForAllPromises(promiseList, [=, this]() {
+            if (abortController->signal().aborted() || !document->isFullyActive() || !m_ongoingNavigateEvent)
+                return;
 
+            m_ongoingNavigateEvent->finish();
             m_ongoingNavigateEvent = nullptr;
 
-            event->finish();
+            dispatchEvent(Event::create(eventNames().navigatesuccessEvent, { }));
 
-            if (!failure) {
-                dispatchEvent(Event::create(eventNames().navigatesuccessEvent, { }));
+            if (RefPtr transition = std::exchange(m_transition, nullptr))
+                transition->resolvePromise();
 
-                if (RefPtr transition = std::exchange(m_transition, nullptr))
-                    transition->resolvePromise();
+            m_ongoingNavigateEvent = nullptr;
 
-                if (apiMethodTracker)
-                    resolveFinishedPromise(apiMethodTracker.get());
-            } else {
-                // FIXME: Fill in error information with exception from promise calls above.
-                auto exception = Exception(ExceptionCode::UnknownError);
-                dispatchEvent(ErrorEvent::create(eventNames().navigateerrorEvent, exception.message(), { }, 0, 0, { }));
+            if (apiMethodTracker)
+                resolveFinishedPromise(apiMethodTracker.get());
 
-                if (RefPtr transition = std::exchange(m_transition, nullptr))
-                    transition->rejectPromise(exception);
-            }
-        } else {
-            // FIXME: and the following failure steps given reason rejectionReason:
+        }, [=, this](JSC::JSValue result) {
+            if (abortController->signal().aborted() || !document->isFullyActive() || !m_ongoingNavigateEvent)
+                return;
+
+            m_ongoingNavigateEvent->finish();
             m_ongoingNavigateEvent = nullptr;
-        }
+
+            // FIXME: Fill in error information.
+            String errorMessage;
+            if (auto* error = jsDynamicCast<JSC::ErrorInstance*>(result))
+                errorMessage = makeString("Uncaught "_s, error->sanitizedToString(protectedScriptExecutionContext()->globalObject()));
+            auto exception = Exception(ExceptionCode::UnknownError, errorMessage);
+
+            dispatchEvent(ErrorEvent::create(eventNames().navigateerrorEvent, exception.message(), document->url().stringWithoutFragmentIdentifier(), 1, 1, { protectedScriptExecutionContext()->globalObject()->vm(), result }));
+
+            if (RefPtr transition = std::exchange(m_transition, nullptr))
+                transition->rejectPromise(exception);
+
+            if (apiMethodTracker) {
+                apiMethodTracker->finishedPromise->rejectWithCallback([result] (auto&) {
```

Comment:
Is it safe to capture result here?  I'm honestly not sure.

Reply:
I think it's fine, copying a JSValue. This callback will only be active when JS is too.